File: OAuthDemoObjc/OAuthDemoObjc/AuthenticationManager.m
//
//  AuthenticationManager.m
//  OAuthDemoObjc
//
//  Created by Hanguang on 1/16/16.
//  Copyright © 2016 Hanguang. All rights reserved.
//

#import "AuthenticationManager.h"


NSString * const baseURL = @"http://integration.flirten.de/api_integration.php/v2/";
//@"http://red.tianwen.flirten.lab/api_dev.php/v2/";
NSString * const oauthURL = @"auth/access_token";
NSString * const clientID = @"flirten";
NSString * const secret = @"secret";
NSString * const scope = @"";
NSString * const serviceProviderIdentifier = @"com.ideawise.koko";
NSString * const myProfileURL = @"account/me";

@interface AuthenticationManager ()
@property (nonatomic, strong) AFOAuth2Manager *oauthManager;
@property (nonatomic, strong) AFHTTPRequestOperationManager *requestManager;
@property (nonatomic, strong) AFOAuthCredential *credential;

@end

@implementation AuthenticationManager

//+ (instancetype)sharedInstance {
//    static AuthenticationManager *_sharedIntance = nil;
//    static dispatch_once_t onceToken;
//    dispatch_once(&onceToken, ^{
//        _sharedIntance = [[self alloc] initPrivate];
//    });
//    
//    return _sharedIntance;
//}
//
//- (instancetype)init NS_UNAVAILABLE
//{
//    return nil;
//}
//
//+ (instancetype)new NS_UNAVAILABLE
//{
//    return nil;
//}

- (instancetype)init {
    self = [super init];
    
    if (self == nil) {
        return nil;
    }
    _requestManager = [[AFHTTPRequestOperationManager alloc] initWithBaseURL:[NSURL URLWithString:baseURL]];
    if (self.credential) {
        [_requestManager.requestSerializer setAuthorizationHeaderFieldWithCredential:self.credential];
    }
    _oauthManager = [[AFOAuth2Manager alloc] initWithBaseURL:[NSURL URLWithString:baseURL] clientID:clientID secret:secret];
    
    return self;
}

- (AFOAuthCredential *)credential {
    return [AFOAuthCredential retrieveCredentialWithIdentifier:serviceProviderIdentifier];
}

#pragma mark - Request helper
- (RACSignal *)refreshToken {
    @weakify(self)
    return [RACSignal createSignal:^RACDisposable *(id<RACSubscriber> subscriber) {
        @strongify(self)
        AFHTTPRequestOperation *op =
        [self.oauthManager authenticateUsingOAuthWithURLString:oauthURL
                                                  refreshToken:self.credential.refreshToken
                                                       success:^(AFOAuthCredential *credential) {
                                                           [self updateCredential:credential];
                                                           [subscriber sendNext:credential];
                                                           [subscriber sendCompleted];
                                                       } failure:^(NSError *error) {
                                                           [subscriber sendError:error];
                                                       }];
        return [RACDisposable disposableWithBlock:^{
            [op cancel];
        }];
    }];
}

- (void)updateCredential:(AFOAuthCredential *)cred {
    [self logWithString:[NSString stringWithFormat:@"Credential: %@", cred]];
    [AFOAuthCredential storeCredential:cred withIdentifier:serviceProviderIdentifier];
    [self.requestManager.requestSerializer setAuthorizationHeaderFieldWithCredential:cred];
}

- (RACSignal *)getSignalWithPath:(NSString *)path parameters:(NSDictionary *)parameters {
    return [RACSignal createSignal:^RACDisposable *(id<RACSubscriber> subscriber) {
        AFHTTPRequestOperation *op =
        [self.requestManager GET:path
                      parameters:parameters
                         success:^(AFHTTPRequestOperation * _Nonnull operation, id  _Nonnull responseObject) {
                             [subscriber sendNext:responseObject];
                             [subscriber sendCompleted];
                         } failure:^(AFHTTPRequestOperation * _Nullable operation, NSError * _Nonnull error) {
                             [subscriber sendError:error];
                         }];
        return [RACDisposable disposableWithBlock:^{
            [op cancel];
        }];
    }];
}

- (RACSignal *)postSignalWithPath:(NSString *)path parameters:(NSDictionary *)parameters {
    return [RACSignal createSignal:^RACDisposable *(id<RACSubscriber> subscriber) {
        AFHTTPRequestOperation *op =
        [self.requestManager POST:path
                       parameters:parameters
                          success:^(AFHTTPRequestOperation * _Nonnull operation, id  _Nonnull responseObject) {
                              [subscriber sendNext:responseObject];
                              [subscriber sendCompleted];
                          } failure:^(AFHTTPRequestOperation * _Nullable operation, NSError * _Nonnull error) {
                              [subscriber sendError:error];
                          }];
        return [RACDisposable disposableWithBlock:^{
            [op cancel];
        }];
    }];
}

- (RACSignal *)refreshTokenIfNeededWithSignal:(RACSignal *)requestSignal {
    return [requestSignal catch:^RACSignal *(NSError *error) {
        if (error) {
            BOOL hasRefreshToken = self.credential.refreshToken != nil;
            BOOL httpCode401AccessDenied = error.code == -1011;
            if (hasRefreshToken && httpCode401AccessDenied) {
                return [[[self refreshToken] ignoreValues] concat:requestSignal];
            }
            return [RACSignal error:error];
        } else {
            return requestSignal;
        }
    }];
}

#pragma mark - Request API

- (RACSignal *)requestForAccessToken {
    return [RACSignal createSignal:^RACDisposable *(id<RACSubscriber> subscriber) {
        AFHTTPRequestOperation *op =
        [self.oauthManager authenticateUsingOAuthWithURLString:oauthURL
                                                         scope:@""
                                                       success:^(AFOAuthCredential *credential) {
                                                           [self updateCredential:credential];
                                                           [subscriber sendNext:credential];
                                                           [subscriber sendCompleted];
                                                       }
                                                       failure:^(NSError *error) {
                                                           [self logWithString:[NSString
                                                                                stringWithFormat:@"Error: %@",
                                                                                error.localizedDescription]];
                                                           [subscriber sendError:error];
                                                       }];
        return [RACDisposable disposableWithBlock:^{
            [op cancel];
        }];

    }];
    
}

- (RACSignal *)requestForOAuthWithUsername:(NSString *)username password:(NSString *)password {
    return [RACSignal createSignal:^RACDisposable *(id<RACSubscriber> subscriber) {
        NSDictionary *parameters = @{@"client_id" : clientID,
                                     @"client_secret" : secret,
                                     @"grant_type" : kAFOAuthPasswordCredentialsGrantType,
                                     @"username" : username,
                                     @"password" : password};
        AFHTTPRequestOperation *op =
        [self.requestManager POST:oauthURL
                       parameters:parameters
                          success:^(AFHTTPRequestOperation * _Nonnull operation, id  _Nonnull responseObject) {
                              if (responseObject) {
                                  NSString *accessToken = responseObject[@"access_token"];
                                  NSString *refreshToken = responseObject[@"refresh_token"];
                                  NSString *tokenType = responseObject[@"token_type"];
                                  NSNumber *expiresIn = responseObject[@"expires_in"];
                                  NSDate *expireDate = [NSDate distantFuture];
                                  
                                  if (expiresIn) {
                                      expireDate = [NSDate dateWithTimeIntervalSinceNow:expiresIn.doubleValue];
                                  }
                                  
                                  AFOAuthCredential *cred = [AFOAuthCredential credentialWithOAuthToken:accessToken tokenType:tokenType];
                                  [cred setExpiration:expireDate];
                                  [cred setRefreshToken:refreshToken];
                                  [self updateCredential:cred];
                                  
                                  [self logWithString:[NSString stringWithFormat:@"Response object: %@", responseObject]];
                                  [subscriber sendNext:responseObject];
                                  [subscriber sendCompleted];
                              }
                          } failure:^(AFHTTPRequestOperation * _Nullable operation, NSError * _Nonnull error) {
                              [self logWithString:[NSString stringWithFormat:@"Error: %@", error]];
                              [subscriber sendError:error];
                          }];
        return [RACDisposable disposableWithBlock:^{
            [op cancel];
        }];
    }];
}

- (RACSignal *)getMyProfile {
    RACSignal *requestSignal = [[[self getSignalWithPath:myProfileURL parameters:nil]
                                deliverOnMainThread]
                                map:^id(NSDictionary *responseObject) {
                                    [self logWithString:[NSString stringWithFormat:@"%@", responseObject]];
                                    return responseObject;
                                }];
    return [self refreshTokenIfNeededWithSignal:requestSignal];
}

#pragma mark - Log
- (void)logWithString:(NSString *)log {
    NSNotification *noti = [NSNotification notificationWithName:AFNetworkActivityLoggerNotification object:log];
    [[NSNotificationCenter defaultCenter] postNotification:noti];
}

@end

File: OAuth2Demo/OAuth2Demo/AuthenticationManager.swift
//
//  AuthenticationManager.swift
//  OAuth2Test
//
//  Created by Hanguang on 1/15/16.
//  Copyright © 2016 Hanguang. All rights reserved.
//

import Cocoa
import AFNetworking
import ReactiveCocoa

class AuthenticationManager {
    let baseURL = "http://integration.flirten.de/api_integration.php/"
    let oauthURL = "v2/auth/access_token"
    let clientID = "flirten"
    let secret = "secret"
    let scope = ""
    let serviceProviderIdentifier = "com.ideawise.koko"
    
    // MARK: - API URL
    let myProfileURL = "v2/account/me"
    
    var credential: AFOAuthCredential {
        get {
            return AFOAuthCredential.retrieveCredentialWithIdentifier(serviceProviderIdentifier)
        }
    }
    
    let oauthManager: AFOAuth2Manager
    let requestManager: AFHTTPRequestOperationManager
    
    
    init() {
        requestManager = AFHTTPRequestOperationManager(baseURL: NSURL(string: baseURL))
        oauthManager = AFOAuth2Manager(baseURL: NSURL(string: baseURL), clientID: clientID, secret: secret)
    }
    
    
    // MARK: - Request method
    func requestOAuthWithSuccess(success: (cred: AFOAuthCredential) -> Void,
        failure: (error: NSError) -> Void) {
            oauthManager.authenticateUsingOAuthWithURLString(oauthURL,
                scope: "",
                success: { (cred) -> Void in
                    self.updateCredential(cred)
                    success(cred: cred)
                }) { (error) -> Void in
                    let log = "Error: \(error.localizedDescription)"
                    self.logWithString(log)
            }
    }
    
    func requestOAuthWith(username: String, password: String,
        success: (op: AFHTTPRequestOperation?, responseObject: AnyObject?) -> Void,
        failure: (op: AFHTTPRequestOperation?, error: NSError?) -> Void) {
            let parameters = ["client_id" : clientID,
                "client_secret" : secret,
                "grant_type" : "password",
                "username" : username,
                "password" : password]
            
            requestManager.POST(oauthURL,
                parameters: parameters,
                success: { (op, responseObject) -> Void in
                    let dict = responseObject as! Dictionary<String, AnyObject>
                    let accessToken = dict["access_token"] as! String
                    let refreshToken = dict["refresh_token"] as! String
                    let tokenType = dict["token_type"] as! String
                    // Expiration is optional, but recommended in the OAuth2 spec. It not provide, assume distantFuture === never expires
                    var expireDate = NSDate.distantFuture()
                    let expiresIn = dict["expires_in"] as? Double
                    if let expiresIn = expiresIn {
                        expireDate = NSDate.init(timeIntervalSinceNow: expiresIn)
                    }
                    
                    let cred = AFOAuthCredential.init(OAuthToken: accessToken, tokenType: tokenType)
                    cred.setExpiration(expireDate)
                    cred.setRefreshToken(refreshToken)
                    self.updateCredential(cred)
                    
                    let log = "ResponseObject: \(responseObject)"
                    self.logWithString(log)
                    
                    success(op: op, responseObject: responseObject)
                }) { (op, error) -> Void in
                    let log = ("Error: \(error.localizedDescription)")
                    self.logWithString(log)
                    failure(op: op, error: error)
            }
    }
    
    func refreshToken() {
        oauthManager.authenticateUsingOAuthWithURLString(oauthURL,
            refreshToken: credential.refreshToken,
            success: { (cred) -> Void in
                self.logWithString("Updated crendential: \(cred)")
                self.updateCredential(cred)
            }) { (error) -> Void in
                print("Error: \(error.localizedDescription)")
        }
    }
    
    func updateCredential(cred: AFOAuthCredential) {
        AFOAuthCredential.storeCredential(cred, withIdentifier: self.serviceProviderIdentifier)
        requestManager.requestSerializer.setAuthorizationHeaderFieldWithCredential(cred)
    }
    
    func requestGetWithPath(path: String, parameters: [String : AnyObject]) -> Signal<AFHTTPRequestOperation?, NSError> {
        return RACSignal.createSignal({ (subscriber) -> RACDisposable! in
            let op = self.requestManager.GET(path, parameters: parameters,
                success: { (operation, responseObject) -> Void in
                    let tuple = RACTuple(objectsFromArray: [operation, responseObject])
                    subscriber.sendNext(tuple)
                    subscriber.sendCompleted()
                },
                failure: { (operation, error) -> Void in
                    subscriber.sendError(error)
            })
            
            return RACDisposable(block: { () -> Void in
                op?.cancel()
            })
        })
    }
    
    func requestPostWithPath(path: String, parameters: [String : AnyObject]) -> RACSignal {
        return RACSignal.createSignal({ (subscriber) -> RACDisposable! in
            let op = self.requestManager.POST(path, parameters: parameters,
                success: { (operation, responseObject) -> Void in
                    let tuple = RACTuple(objectsFromArray: [operation, responseObject])
                    subscriber.sendNext(tuple)
                    subscriber.sendCompleted()
                },
                failure: { (operation, error) -> Void in
                    subscriber.sendError(error)
            })
            
            return RACDisposable(block: { () -> Void in
                op?.cancel()
            })
        })
    }
    
    func refreshTokenIfNecessaryWithSignal(signal: RACSignal) -> RACSignal {
    }
    
    func getMyProfile() -> RACSignal {
        let requestSignal = requestGetWithPath(myProfileURL, parameters: [:])
        requestSignal.map { (tuple) -> AnyObject! in
            let response = tuple as! RACTuple
            let dict = response.second
            self.logWithString("Response object: \(dict)")
            return dict
        }
        
        return refreshTokenIfNecessaryWithSignal(requestSignal)
    }
    
    // MARK: - LOG
    func logWithString(log: String) {
        let notification = NSNotification(name: "AFNetworkActivityLoggerNotification", object: log)
        NSNotificationCenter.defaultCenter().postNotification(notification)
    }
}


